### ISSUE TITLE
When decodeFromJsonElement, enum parameters with no default value throw NoSuchElementException

### ISSUE THREAD

**Issue by dahyun1226** [Kotlin/kotlinx.serialization#2222]
**Describe the bug**
The decodeFromJsonElement function uses the JsonTreeDecoder when element is a JsonObject.
However, the decodeElementIndex function in the JsonTreeDecoder can reach the coerceInputValue function with the explicitNulls = false option and coerceInputValues = true option provided by Json, and the situation of enum parameters with no default value.

<img width="600" alt="image" src="https://user-images.githubusercontent.com/60708291/223621233-af2c8732-e343-45f8-b14a-2038983aa938.png">
In coerceInputValue function, currentElement(tag) can throw NoSuchElementException because currentElement assume that value with tag exist.
<img width="600" alt="image" src="https://user-images.githubusercontent.com/60708291/223621341-55e91d8d-20f1-4a3b-962b-3eacf87deb7e.png">
<img width="600" alt="image" src="https://user-images.githubusercontent.com/60708291/223621370-3f92c748-8e62-40aa-b7f3-07a8047ee98b.png">
I don't think this was intended.

**To Reproduce**
It's a bit of a contrived situation test, but I think it's good enough to reproduce the situation.
The following tests will fail.
```kotlin
    @Serializable
    enum class TestEnum {
        ENUM_A
    }

    @OptIn(ExperimentalSerializationApi::class)
    private val json = Json {
        prettyPrint = false
        ignoreUnknownKeys = true
        encodeDefaults = true
        explicitNulls = false
        coerceInputValues = true
    }

    object TestClassSerializer : KSerializer<TestClass> {
        override val descriptor: SerialDescriptor
            get() = TestClassSurrogate.serializer().descriptor

        override fun deserialize(decoder: Decoder): TestClass {
            val surrogate = decoder.decodeSerializableValue(TestClassSurrogate.serializer())
            return TestClass(testEnum = surrogate.testEnum)
        }

        override fun serialize(encoder: Encoder, value: TestClass) {}
    }

    @Serializable(with = TestClassSerializer::class)
    data class TestClass(val testEnum: TestEnum?)

    @Serializable
    data class TestClassSurrogate(val testEnum: TestEnum?)

    @Test
    fun test() {
        val jsonString = """{}""".trimIndent()

        val jsonElement = json.decodeFromString(JsonElement.serializer(), jsonString)

        val actualTestClass = json.decodeFromJsonElement<TestClass>(jsonElement)

        Truth.assertThat(actualTestClass.testEnum).isNull()
    }
```
**Environment**
 - Kotlin version: [1.7.10]
 - Library version: [1.4.1]
 - Kotlin platforms: [JVM]
 - Gradle version: [7.4]
 - IDE version (if bug is related to the IDE) [Android Studio Electric Eel 2022.1.1 Patch 2]

**Comment by sandwwraith** [comment]
Please check 1.5.0 — it has fix for https://github.com/Kotlin/kotlinx.serialization/issues/2170 which looks similar

**Comment by dahyun1226** [comment]
The error still occurs in version 1.5.0.
It is still the same to use currentElement when accessing peekNull, and this is where the error occurs.

**Comment by sandwwraith** [comment]
I've checked on the current `dev`, and the issue is not present anymore. It was likely fixed by #2530. Also, linking #2586 as a reference — it may be related.
